### PR TITLE
Integrate 2023-12-01 API Preview Changes

### PR DESCRIPTION
**Pull request by timstewartm**
Updates SDK changes to work with new preview SDK given provided swaggers.

### CODE REVIEW

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 1]
```diff
@@ -65,7 +65,7 @@ public class DiscoveryDeviceClient
         /// </summary>
         /// <param name="cancellationToken"></param>
         /// <returns>nonce for authentication challenge</returns>
-        public Task<string> IssueChallengeAsync(CancellationToken cancellationToken = default)
+        public Task<byte[]> IssueChallengeAsync(CancellationToken cancellationToken = default)
```
Comment (deepanjyoti-sarkar) [inline discovery/device/src/DiscoveryDeviceClient.cs:68]: I believe the contract is IssueChallengeAsync gives a nonce which is then later passed along the GetOnboardingInfoAsync call? Is that right?
If yes, do the public signature of these two methods need to change? Is there any interpretation of the nonce in the consumer of the SDK? If the signature changes here, then the device code (which consumes the SDK) also needs to change to re-compile.

File: discovery/device/src/DiscoveryTransportGetOnboardingInfoRequest.cs [hunk 2]
```diff
@@ -13,15 +13,15 @@ public class DiscoveryTransportGetOnboardingInfoRequest : DiscoveryTransportRequ
...
         /// </summary>
-        public string Nonce { get; private set; }
+        public byte[] Nonce { get; private set; }
 
         /// <summary>
         /// 
         /// </summary>
         /// <param name="globalDeviceEndpoint"></param>
         /// <param name="security"></param>
         /// <param name="nonce"></param>
```
Comment (deepanjyoti-sarkar) [inline discovery/device/src/DiscoveryTransportGetOnboardingInfoRequest.cs]: nit : Please file the details around the params.

File: discovery/device/src/DiscoveryDeviceClient.cs (hunk 1 above)
Comment (deepanjyoti-sarkar) [reply discovery/device/src/DiscoveryDeviceClient.cs:68]: I'm ok with the change from string to byte[] as such, but want to ensure that we are conscious about what changes results in the change in the consumer code (which the SDK layer is normally supposed to abstract out)

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 3]
```diff
@@ -85,7 +85,7 @@ public Task<string> IssueChallengeAsync(CancellationToken cancellationToken = de
         /// <param name="nonce"></param>
```
Comment (deepanjyoti-sarkar) [inline discovery/device/src/DiscoveryDeviceClient.cs]: nit: Please fill up the param with one liners.

File: discovery/device/src/OnboardingInfo.cs [hunk 4]
```diff
@@ -18,14 +18,14 @@ public class OnboardingInfo
         /// <summary>
         /// Credentials to authenticate with the provisioning service
         /// </summary>
-        public X509Certificate2 ProvisioningCertificate { get; private set; }
+        public X509Certificate2Collection ProvisioningCertificate { get; private set; }
```
Comment (deepanjyoti-sarkar) [inline discovery/device/src/OnboardingInfo.cs:21]: Why this change?

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 5]
```diff
@@ -14,6 +14,7 @@
 using Microsoft.Azure.Devices.Shared;
 using Newtonsoft.Json;
 using System.Text;
+using System.Collections.Generic;
```
Comment (deepanjyoti-sarkar) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs]: nit: remove and sort usings.

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 6]
```diff
@@ -72,7 +73,7 @@ public override async Task<string> IssueChallengeAsync(DiscoveryTransportIssueCh
                     Port = Port,
                 };
 
-                using EdgeDiscoveryService client = new EdgeDiscoveryService(builder.Uri, httpClientHandler);
+                using MicrosoftFairfieldGardensDiscovery client = new MicrosoftFairfieldGardensDiscovery(builder.Uri, httpClientHandler);
```
Comment (deepanjyoti-sarkar) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs:75]: use the interface IMicrosoftFairfieldGardensDiscovery here.

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 7]
```diff
@@ -81,11 +82,11 @@ public override async Task<string> IssueChallengeAsync(DiscoveryTransportIssueCh
...
                 var onboardRequest = new ChallengeRequest(
                     registrationId, 
-                    Convert.ToBase64String(securityProvider.GetEndorsementKey()), 
-                    Convert.ToBase64String(securityProvider.GetStorageRootKey()));
+                    securityProvider.GetEndorsementKey(), 
+                    securityProvider.GetStorageRootKey());
 
                 Challenge challenge = await client.DiscoveryRegistrations
-                    .IssueChallengeAsync(onboardRequest, cancellationToken)
+                    .IssueChallengeAsync("2023-12-01-preview", onboardRequest, cancellationToken: cancellationToken)
```
Comment (deepanjyoti-sarkar) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs]: define a const with the apiVersion = "2023-12-01-preview". Assuming I read that right...

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 8]
```diff
@@ -180,7 +181,7 @@ public override async Task<OnboardingInfo> GetOnboardingInfoAsync(DiscoveryTrans
                     Port = Port,
                 };
 
-                using EdgeDiscoveryService client = new EdgeDiscoveryService(builder.Uri, httpClientHandler);
+                using MicrosoftFairfieldGardensDiscovery client = new MicrosoftFairfieldGardensDiscovery(builder.Uri, httpClientHandler);
```
Comment (deepanjyoti-sarkar) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs:183]: use inteface IMicrosoftFairfieldGardensDiscovery here

File: discovery/device/src/OnboardingInfo.cs (hunk 4 above)
Comment (timstewartm) [reply discovery/device/src/OnboardingInfo.cs:21]: The API was changed from giving us just one certificate to provided the entire cert chain. The client facing code was changed to reflect this.

File: discovery/device/src/DiscoveryDeviceClient.cs (hunk 1 above)
Comment (timstewartm) [reply discovery/device/src/DiscoveryDeviceClient.cs:68]: > I believe the contract is IssueChallengeAsync gives a nonce which is then later passed along the GetOnboardingInfoAsync call? Is that right?

Yes that is correct

There is no client interpretation of the nonce but this change is cleaner overall since the alternative would be converting to string, then converting from string back to byte array within the SDK.

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs (hunk 8 above)
Comment (timstewartm) [reply discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs:183]: The interface does not expose the HttpClient, which we need access to.

File: discovery/device/src/DiscoveryDeviceClient.cs (hunk 1 above)
Comment (deepanjyoti-sarkar) [reply discovery/device/src/DiscoveryDeviceClient.cs:68]: Yes, I was ok with the change given that we haven't shipped anything yet... However these are areas to keep in mind once things go out in the field.

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 9]
```diff
@@ -205,37 +199,38 @@ public override async Task<OnboardingInfo> GetOnboardingInfoAsync(DiscoveryTrans
...
+                var onboardInfoRequest = new BootstrapRequest(
+                    registrationId,
+                    securityProvider.GetEndorsementKey(),
+                    securityProvider.GetStorageRootKey(),
+                    csr);
 
                 BootstrapResponse onboardInfo = await client.DiscoveryRegistrations
-                    .GetOnboardingInfoAsync(onboardInfoRequest, sasToken, cancellationToken)
+                    .GetOnboardingInfoAsync(Versions.V20231201Preview, onboardInfoRequest, sasToken, cancellationToken: cancellationToken)
                     .ConfigureAwait(false);
```
Comment (deepanjyoti-sarkar) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs:212]: PLease make a constant for this api version value...

**Review by deepanjyoti-sarkar — APPROVED** [review]
